### DISCUSSION TITLE
Rendering a loader while SSR is in progress

### DISCUSSION THREAD

**Discussion by Shooshte** [vercel/next.js#41324] [Help]
I would like to show a loader component while server side rendering (SSR) is in progress. 
My current solution includes:
- subscribing to router `routeChangeStart`, `routeChangeComplete` and `routeChangeError` events in my `_app.tsx`
- toggling a state when a route is change if it matches a certain pattern (a hardcoded array of strings matching my SSR routes)
- showing a loader in the rendered component if the SSR state is `true`

This is the `useEffect` function:
```javascript
  useEffect(() => {
    const handleRouteChange = (url: string) => {
    const SSR_ROUTES = ['/recipes'];

      if (SSR_ROUTES.includes(url)) {
        setSSRInProgress(true);
      }
    };

    const handleRouteChangeComplete = () => {
      setSSRInProgress(false);
    };

    router.events.on('routeChangeStart', handleRouteChange);
    router.events.on('routeChangeComplete', handleRouteChangeComplete);
    router.events.on('routeChangeError', handleRouteChangeComplete);

    return () => {
      router.events.off('routeChangeStart', handleRouteChange);
      router.events.off('routeChangeComplete', handleRouteChangeComplete);
      router.events.off('routeChangeError', handleRouteChangeComplete);
    };

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  ```
  
  While this works, there are a couple of issues:
  - The number of SSR routes will grow, which might make the pattern matching of my routes slow
  - The loader is rendered twice; first in the component that initiates the change, then in the component the new route mounts on first render, before `routeChangeComplete` triggers.
  
 Does anyone know of a better way to handle this? A solution would be to render a loader inside the `_app.tsx` page, but I want to avoid that as my page components show Loaders inside different parts of their UI.

**Comment by icyJoseph** [comment]
Usually you kinda want to attach listeners to the Router instance, and you can do that in `_app`, but not necessarily inside the App React component, but at the top of the file.

```js
import type { AppProps } from "next/app";
import Router from "next/router";

Router.events.on("routeChangeStart", (path) => {
  console.log("routeChangeStart", { path });
});

Router.events.on("routeChangeComplete", (path) => {
  console.log("routeChangeStart", { path });
});

Router.events.on("routeChangeError", (path) => {
  console.log("routeChangeError", { path });
});

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

export default MyApp;

```

A popular library to show a spinner in this case is [`nprogress`](https://ricostacruz.com/nprogress/), and of course there's a family of variations. The key concept would be that you control DOM elements external to the React rendered DOM via your library. Or that you mount a React component to the DOM and have it listen to external events to trigger behavior, like [react-toastify](https://fkhadra.github.io/react-toastify/installation) does.
